docs/adr/0001-update-logins-api.md

```diff
@@ -0,0 +1,77 @@
+* Establishing a standard encyrption approach across components
+
+
+## Considered Options
+
+* Option 1 - Reduce the API functions that require the encryption key and pass the key the remaining functions
+* Option 2 - Implement a different encryption approach
+
+
+## Decision Outcome
+
+(I'm purposely leaving this blank until we arrive at a decision.)
+
+
+## Pros and Cons of the Options
+
+### Option 1 - Reduce the API functions that require the encryption key and pass the key the remaining functions
+
+* Description
+
+  These are the logins API functions that will require the per-field encryption key:
+    - [add](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L362)
+    - [update](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L611)
+    - [get_all](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L273)
+    - [get_by_base_domain](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L279)
+    - [get_by_id](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L330)
+    - [check_valid_with_no_dupes](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L663)
+    - [potential_dupes_ignoring_username](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L721)
+    - [import_multiple](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/db.rs#L455)
+
+      **Note:** Functions related to sync have been omitted as it’s assumed they’ll have access to decrypted data.
+
+  The proposed changes will reduce the above list of functions to the following:
+    - `save_or_update`
+      - This function will replace `add` and `update`
+    - `import_multiple`
+      - This function will be removed once the Fennec to Fenix migration period ends
+
+  Both the `potential_dupes_ignoring_username` and `check_valid_with_no_dupes` functions will removed from the API. Neither function is called in Firefox iOS. Android Components uses both to provide validation and de-duping before logins are added or updated. We can eliminate the need to externalize these functions by replicating this logic in the new `save_or_update` function.
+
+  If we exclude the encrypted `username` and `password` fields from [the Nul check](https://github.com/mozilla/application-services/blob/1248a352cb2701b92395f2783bee8a88d18de455/components/logins/src/login.rs#L446) in the validate and fixup logic, we can eliminate the need for the `get_all`, `get_by_base_domain`, and `get_by_id` functions to have the encryption key.
```

Comment:
I just want to highlight this part of the proposal because this is where I am uncertain. At present, the three "get" API functions I mentioned above call `Login::from_row` which calls `maybe_fixup` which calls `validate_and_fixup`, which is where the `Nul` check occurs. My understanding is that we're fixing up logins upon db retrieval in order to gradually clean up our login records. If we want to continue to do this, I'm choosing to exclude the encrypted password and username fields so that the get functions don't need the encryption key. But if we no longer need to fixup retrieved records, that simplifies things. But for certain if we want to leave the validation logic unchanged and continue fixing up retrieved records, all three get functions will need the encryption key.

Reply:
I don't have all the context from design discussions in the autofill component, so the below may not be tractable, but I wanted to suggest it for completeness.

IIUC, the proposed API exposes `encrypt_string` and `decrypt_string` functions that consumers need to use in order to access the individual encrypted fields. Would it be useful to instead have a higher-level `decrypt_login_record` function that takes a whole login record with encrypted fields, and:

* If the encrypted fields are all valid, returns some sort of `PlainTextLogin` object will all fields decrypted
* If the encrypted fields contained invalid data, repairs the data in the db and then returns an error

I kind of feel like we already decided not to do this in the autofill component, in which case, perhaps it's worth us recording the reasons why as part of this ADR.

Reply:
> Would it be useful to instead have a higher-level `decrypt_login_record`

Possibly, yeah - especially here where we have 2 encrypted fields. However, unless I misunderstand, this doesn't actually solve the problems here - it would still require the "get" functions to have the key available, which I think we should try and avoid. None of our other components do these kinds of fixups at "get" time. I'd be fine with a new API designed expressly for fixups that did require the key, but I'm inclined to suggest the API shape is wrong where getters end up writing fixed up records to the DB.

Reply:
In a little more detail, was I was thinking here was something like the following consumer experience:

```
let encryptedLogin = store.get_by_id("foobar")
try:
  let decryptedLogin = store.decrypt_login_record(key, encryptedLogin)
except CorruptedLoginRecordError:
  print("oh no what do we do?")
```

That is, guard against invalid data at the point where you decrypt the fields.

Having written it out like that though, I'm not sure I like it. On one hand, it seems like it could spread the "fixup at read time" logic to lots of parts of the *consumer* code, which architecturally seems like the opposite of what we want. On the other hand, it seems better to refuse to decrypt corrupted data such as a login with a nul in the username, rather than hand out known-invalid data to the consumer.

Reply:
Actually, I think this may work?? I mean if these three "get" functions didn't call the fixup logic, the consumer would retrieve the encrypted data and then be responsible for decrypting it. That's the pattern that was established in autofill anyway. So if we have this higher level `decrypt_login_record` function, we'd be adhering to the same pattern. And then we could remove the fixup logic from the get calls.